File: precast-crm/src/lib/dashboard-metrics.ts
/**
 * Pure arithmetic + bucketing helpers behind the dashboard's money metrics.
 * Deliberately Prisma-free so the maths is unit-testable (see
 * `dashboard-metrics.test.ts`).
 *
 * ── The two money metrics ──────────────────────────────────────────────
 * The dashboard used to ship two different numbers both labelled "revenue".
 * They are NOT interchangeable and neither may be called bare
 * "revenue" / "даромад" anywhere in the code or the UI:
 *
 *   • BOOKED    «Буюртма қилинган · Booked»
 *     Σ Order.totalPrice over live orders, bucketed by `Order.placedAt`.
 *     What the business SOLD in a period.
 *
 *   • COLLECTED «Тушган пул · Collected»
 *     Σ Payment.amount over CONFIRMED payments, bucketed by
 *     `Payment.confirmedAt`. What the business actually RECEIVED.
 *
 * Booked − Collected is (roughly) the outstanding receivable, which is why
 * the two numbers never agreed.
 *
 * `Order.confirmedPaid` is a running total with NO date attached, so it can
 * never answer "how much cash arrived in May" — any collections time series
 * has to come off the Payment table.
 *
 * ── Timezone ──────────────────────────────────────────────────────────
 * Every bucket key is built from the server's LOCAL calendar fields
 * (`getFullYear`/`getMonth`/`getDate`). Production runs with
 * `TZ=Asia/Tashkent`; local dates are stored as the previous day 19:00Z, so
 * UTC bucketing / `date_trunc` would shift every date by one day.
 */

/** Short Uzbek Cyrillic month names, indexed by `Date#getMonth()`. */
export const MONTH_UZ_SHORT = [
  'Янв', 'Фев', 'Мар', 'Апр', 'Май', 'Июн',
  'Июл', 'Авг', 'Сен', 'Окт', 'Ноя', 'Дек',
] as const;

/** `YYYY-MM` in the server's local zone. */
export function monthKey(d: Date): string {
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`;
}

/** `YYYY-MM-DD` in the server's local zone. */
export function dayKey(d: Date): string {
  return (
    d.getFullYear() +
    '-' +
    String(d.getMonth() + 1).padStart(2, '0') +
    '-' +
    String(d.getDate()).padStart(2, '0')
  );
}

export interface MonthBucket {
  /** `YYYY-MM` — join key for accumulated sums. */
  key: string;
  /** Uzbek Cyrillic short label for the x-axis. */
  label: string;
  year: number;
  /** 0-indexed, as `Date#getMonth()`. */
  monthIndex: number;
}

/**
 * ── The two date bases ────────────────────────────────────────────────
 * An order carries two meaningful dates that answer different questions:
 *
 *   • ORDER DATE    `Order.placedAt`    — when the business was WON.
 *     IMMUTABLE. A placed order never moves between months, so a figure
 *     bucketed by `placedAt` is final once the month has closed. Nothing
 *     is ever placed in the future, so this basis is purely backward
 *     looking (trailing 12 months).
 *
 *   • DELIVERY DATE `Order.scheduledAt` — when the work is PROMISED.
 *     MUTABLE. Rescheduling moves an order — and its whole value — from
 *     one month into another, so a PAST month's figure on this basis can
 *     still change. That is correct behaviour, but it has to be surfaced
 *     in the UI or the number stops being trustworthy.
 *
 * Bucketing by `placedAt` alone hides committed FUTURE work: an order
 * placed in August and rescheduled into September reads as August money
 * while September reads zero. The delivery basis therefore needs a window
 * that extends forward (`deliveryMonthWindow`) — a trailing-only window
 * would hide exactly the months this basis exists to show.
 */
export type DateBasis = 'order' | 'delivery';

/**
 * Trailing months carried BEFORE the current one on the delivery basis.
 * 8 + current + 3 forward = 12 buckets — deliberately the same width as
 * the order-date window, and the widest span that still yields 12 DISTINCT
 * month labels. A 13-month span would print the same Cyrillic short name
 * twice (Ноя '25 and Ноя '26) on the x-axis and in the month picker.
 */
export const DELIVERY_WINDOW_BACK = 8;
/** Future months carried AFTER the current one on the delivery basis. */
export const DELIVERY_WINDOW_FORWARD = 3;
/** Trailing months, current one included, on the order-date basis. */
export const ORDER_WINDOW_MONTHS = 12;

/**
 * Contiguous month buckets from `back` months before the month holding
 * `now` through `forward` months after it, ordered oldest → newest —
 * `back + 1 + forward` buckets. Empty months still get a bucket so a line
 * chart keeps a continuous x-axis.
 */
export function monthWindowSpan(now: Date, back: number, forward: number): MonthBucket[] {
  const out: MonthBucket[] = [];
  for (let offset = -back; offset <= forward; offset++) {
    const d = new Date(now.getFullYear(), now.getMonth() + offset, 1);
    out.push({
      key: monthKey(d),
      label: MONTH_UZ_SHORT[d.getMonth()]!,
      year: d.getFullYear(),
      monthIndex: d.getMonth(),
    });
  }
  return out;
}

/**
 * `months` contiguous buckets ending with the month that contains `now`,
 * ordered oldest → newest. This is the ORDER-DATE window: backward-only,
 * because nothing is ever placed in the future.
 */
export function monthWindow(now: Date, months: number): MonthBucket[] {
  return monthWindowSpan(now, months - 1, 0);
}

/**
 * The DELIVERY-DATE window: `DELIVERY_WINDOW_BACK` trailing months, the
 * current month, and `DELIVERY_WINDOW_FORWARD` months of committed future
 * work. Forward-looking on purpose — scheduled work sits in months that
 * have not happened yet.
 */
export function deliveryMonthWindow(now: Date): MonthBucket[] {
  return monthWindowSpan(now, DELIVERY_WINDOW_BACK, DELIVERY_WINDOW_FORWARD);
}

/**
 * Index of the bucket holding `now`. On the order-date window that is the
 * LAST index; on the delivery window it sits `DELIVERY_WINDOW_FORWARD`
 * buckets from the end, so "current month" can never be assumed to be
 * `length - 1`. Falls back to the last bucket when the window does not
 * contain `now` at all.
 */
export function currentMonthIndex(buckets: readonly MonthBucket[], now: Date): number {
  const key = monthKey(now);
  const i = buckets.findIndex((b) => b.key === key);
  return i >= 0 ? i : Math.max(buckets.length - 1, 0);
}

/**
 * Translate a selected month index from one window into another, used when
 * the operator flips the date basis so the SAME calendar month stays
 * selected instead of the selection silently jumping to another month.
 * Falls back to clamping the raw index into range when that month is not
 * in the target window — e.g. a future month picked on the delivery basis
 * and then flipped back to the order basis, which carries no future.
 */
export function remapMonthIndex(
  fromKeys: readonly string[],
  toKeys: readonly string[],
  idx: number,
): number {
  if (toKeys.length === 0) return 0;
  const key = fromKeys[idx];
  if (key !== undefined) {
    const found = toKeys.indexOf(key);
    if (found >= 0) return found;
  }
  return Math.min(Math.max(idx, 0), toKeys.length - 1);
}

export interface MonthTotal {
  total: number;
  count: number;
}

/**
 * Sum `getAmount` into local-calendar month buckets. Rows whose date is
 * null are skipped and reported nowhere — e.g. a CONFIRMED payment with no
 * `confirmedAt` cannot be attributed to a month, so it contributes to the
 * all-time collected total but to no bucket.
 */
export function accumulateByMonth<T>(
  rows: readonly T[],
  getDate: (row: T) => Date | null,
  getAmount: (row: T) => number,
): Map<string, MonthTotal> {
  const map = new Map<string, MonthTotal>();
  for (const row of rows) {
    const d = getDate(row);
    if (!d) continue;
    const key = monthKey(d);
    const cur = map.get(key) ?? { total: 0, count: 0 };
    cur.total += getAmount(row);
    cur.count += 1;
    map.set(key, cur);
  }
  return map;
}

/**
 * Average order value = booked value per order.
 *
 * The numerator is Σ `totalPrice` (BOOKED), never Σ `confirmedPaid` — an
 * order that has not been paid yet still has a value, and dividing
 * collections by order count answers a different question entirely
 * ("collections per order"), which is not AOV.
 *
 * Rounded to whole UZS; returns 0 rather than NaN/Infinity on an empty set.
 */
export function averageOrderValue(bookedTotal: number, orderCount: number): number {
  if (orderCount <= 0) return 0;
  return Math.round(bookedTotal / orderCount);
}

/**
 * Outstanding receivable on a single order:
 * `max(0, totalPrice − confirmedPaid − writeOffAmount)`.
 *
 * A write-off flips an order to FULLY_PAID while the cash is short, so
 * `paymentState` is never a proxy for "we received the money" — only this
 * subtraction is.
 */
export function outstandingAmount(row: {
  totalPrice: number;
  confirmedPaid: number;
  writeOffAmount: number;
}): number {
  const due = row.totalPrice - row.confirmedPaid - row.writeOffAmount;
  return due > 0 ? due : 0;
}

/**
 * Number of distinct values. Used to count CUSTOMERS rather than order
 * rows — `groupBy(paymentState)._count.clientId` counts rows, so a client
 * with six orders was previously counted six times under a "customers"
 * label.
 */
export function countDistinct(values: Iterable<string>): number {
  return new Set(values).size;
}

/**
 * Trend pill data. `deltaPct` is rounded to whole percent. `direction`
 * derives from the sign — "flat" when |delta| < 1% so we don't flash a
 * pill for noise. `polarity` controls the green/red color in the UI:
 * "positive" means an up arrow is good (booked/collected), "negative"
 * means an up arrow is bad (receivables).
 */
export interface Trend {
  deltaPct: number;
  direction: 'up' | 'down' | 'flat';
  polarity: 'positive' | 'negative';
}

/**
 * Build a trend pill from a current vs previous-period number pair.
 * Returns null when there's no previous-period basis (the first month
 * of operation, or any case where dividing by zero is meaningless).
 * |delta| < 1% renders as a flat arrow so noise doesn't trigger
 * green/red flashing.
 *
 * Lives here (not in `dashboard-data.ts`) because the dashboard page
 * recomputes these client-side when the operator selects a month other
 * than the current one — the server can only precompute the
 * current-vs-previous pair.
 */
export function buildTrend(
  current: number,
  previous: number,
  polarity: Trend['polarity'],
): Trend | null {
  if (previous <= 0) return null;
  const deltaPct = Math.round(((current - previous) / previous) * 100);
  const direction: Trend['direction'] =
    Math.abs(deltaPct) < 1 ? 'flat' : deltaPct > 0 ? 'up' : 'down';
  return { deltaPct, direction, polarity };
}

File: android/core/model/src/main/kotlin/uz/etalon/crm/core/model/DashboardMetrics.kt
package uz.etalon.crm.core.model

import java.math.BigDecimal
import java.math.RoundingMode

/**
 * The arithmetic `src/lib/dashboard-metrics.ts` does, ported line for line.
 *
 * The dashboard's month picker (design §2.3b) re-scopes the financial rail to any of the twelve
 * months in the payload, and the server can only precompute the current-vs-previous pair — so the
 * phone recomputes the selected month's figures off the series exactly as `FinancialKPIs.tsx`
 * does. "Exactly" is the acceptance criterion: for the CURRENT month these functions must return
 * the very numbers the server already sent in `bookedThisMonth` / `collectedThisMonth` /
 * `averageOrderValue`, and a unit test on the recorded payload asserts that equality.
 *
 * Everything here is `BigDecimal`/[Money]. Money never becomes a `Double` on this path, which is
 * the rule the rest of this module already follows.
 */

/** Scale the divisions are carried at before [jsRound] sees them. Well past the 9 significant
 *  digits a UZS figure has, so the rounding decision is made on the real quotient and not on a
 *  quotient that was already rounded once. */
private const val DIVISION_SCALE = 12

/** `(current − previous) ÷ previous × 100`. */
private val HUNDRED = BigDecimal(100)

/** Half, added before flooring — the whole of JavaScript's rounding rule. */
private val HALF = BigDecimal("0.5")

/**
 * JavaScript's `Math.round`: **floor(x + 0,5)**, which is NOT [RoundingMode.HALF_UP].
 *
 * The two agree on every positive number and disagree on every negative half: `Math.round(-2.5)`
 * is `-2` (towards +∞) while `HALF_UP` gives `-3` (away from zero). Only one figure on this
 * screen can be negative — a trend's `deltaPct` — and it is exactly the figure the server also
 * computes with `Math.round`, so a month that fell by precisely 2,5 % would otherwise read as
 * −3 % on the phone and −2 % on the web for the same data.
 */
fun jsRound(x: BigDecimal): BigDecimal = (x + HALF).setScale(0, RoundingMode.FLOOR)

/**
 * `dashboard-metrics.ts:215` — average order value: booked value per order, in whole UZS.
 *
 * The numerator is Σ `totalPrice` (BOOKED), never Σ collected: an order that has not been paid
 * yet still has a value, and dividing collections by order count answers a different question.
 * Zero rather than a division by zero on an empty (or nonsensical) order set.
 */
fun averageOrderValue(booked: Money, orders: Int): Money {
    if (orders <= 0) return Money.ZERO
    return Money(jsRound(booked.amount.divide(BigDecimal(orders), DIVISION_SCALE, RoundingMode.HALF_UP)))
}

/**
 * `dashboard-metrics.ts:272` — the delta pill for a current-vs-previous pair.
 *
 * `null` when there is no previous-period basis to compare against (the first month of operation,
 * or any case where dividing would be meaningless) — the card then draws no badge at all rather
 * than an invented 0 %. |delta| under one whole percent renders [TrendDirection.FLAT] so noise
 * does not flash a screen full of green and red.
 *
 * [polarity] does not affect the maths; it decides which way the colour runs in the badge (up is
 * good for booked/collected/AOV, bad for receivables).
 */
fun buildTrend(current: Money, previous: Money, polarity: TrendPolarity): Trend? {
    if (previous.amount.signum() <= 0) return null
    val deltaPct = jsRound(
        (current.amount - previous.amount)
            .multiply(HUNDRED)
            .divide(previous.amount, DIVISION_SCALE, RoundingMode.HALF_UP),
    )
    val direction = when {
        deltaPct.abs() < BigDecimal.ONE -> TrendDirection.FLAT
        deltaPct.signum() > 0 -> TrendDirection.UP
        else -> TrendDirection.DOWN
    }
    return Trend(deltaPct = deltaPct, direction = direction, polarity = polarity)
}

/**
 * Everything the financial rail and the «Юкланган ҳажм» tile show for ONE month of the payload's
 * twelve — the phone's half of `FinancialKPIs.tsx` and `dashboard/page.tsx:86-97`.
 *
 * @property idx the month this scope is about, already clamped into the series.
 * @property monthKey `YYYY-MM` — `monthKeys[idx]`, empty when the server sent no keys. It is the
 *   ONLY month identity this scope carries: the screen derives every month name it prints from it
 *   through the app's own formatter, so that casing and the Tashkent fallback are the same rules
 *   every other date follows. The server's own `bookedByMonth[idx].month` label is deliberately
 *   not carried — nothing rendered it, and a second month identity is a second thing to disagree.
 * @property isCurrent whether [idx] is the month containing today — what decides «ушбу ой» against
 *   «{ой} ойи», and the only thing the current month is special about.
 * @property bookedSeries the months ENDING at [idx] — at most [SPARKLINE_MONTHS], fewer when the
 *   account (or the selection) is younger than that. The sparkline left-pads what it is short of,
 *   so the three rail cards keep the same bar count and the same height either way.
 * @property loaded the `loadedVolumeByMonth` row for [monthKey], found by KEY and not by index
 *   (loading has its own date axis), or `null` when that month loaded nothing.
 */
data class MonthScope(
    val idx: Int,
    val monthKey: String,
    val isCurrent: Boolean,
    val booked: PeriodMoney,
    val collected: PeriodMoney,
    val aov: Aov,
    val bookedSeries: List<Money>,
    val collectedSeries: List<Money>,
    val aovSeries: List<Money>,
    val loaded: LoadedVolume?,
)

/** How many months a rail sparkline draws (design §2.3). The web draws six; the phone's rail was
 *  built on eight in Task 3 and the bar geometry is the design system's, so the window stays
 *  eight — the arithmetic inside it is the web's unchanged. */
const val SPARKLINE_MONTHS = 8

/**
 * `[from]..[toInclusive]` of a series, clipped to what the list actually holds.
 *
 * The three series arrive the same length today, and the one that did not would otherwise throw
 * `IndexOutOfBounds` inside a `subList` on a screen that is only drawing a sparkline. Clipping
 * short is the honest failure: fewer bars, never a crash and never a borrowed month.
 */
private inline fun <T, R> List<T>.window(from: Int, toInclusive: Int, pick: (T) -> R): List<R> {
    val lo = from.coerceIn(0, size)
    val hi = (toInclusive + 1).coerceIn(lo, size)
    return subList(lo, hi).map(pick)
}

/**
 * The selected month's figures, read off the twelve-month series the server sent.
 *
 * [idx] is clamped into the series the way `dashboard/page.tsx` clamps it, so a stale selection
 * surviving a refresh that shortened the window can never index past its end. Every delta compares
 * the selected month with the one BEFORE it — not always this month with last month — which is
 * what makes the badges mean something when a past month is picked.
 *
 * For the current month this returns the server's own `bookedThisMonth` / `collectedThisMonth` /
 * `averageOrderValue` figures, because the server derived them from the same rows with the same
 * two functions. That equality is the parity guard and is asserted against a recorded payload.
 *
 * **The trends are recomputed here, and the payload's own `trend` fields are deliberately never
 * rendered.** The server derives its trend from a database aggregate of the previous *calendar*
 * month (`dashboard-data.ts:487-499`), which can only ever answer "this month vs last month"; a
 * month picker needs "the selected month vs the one before it", which only the series can answer.
 * `FinancialKPIs.tsx:121-124` substitutes exactly the same way — it ignores the precomputed trends
 * and calls `buildTrend` on the series for every month including the current one — so recomputing
 * is what matches the web, which is the acceptance criterion. The two agree on the current month
 * whenever the previous calendar month is inside the window, which is every month the phone can
 * select; `PeriodMoney.trend` and [Aov.trend] stay on the model for exactly that comparison, and
 * the parity guard makes it.
 */
fun monthScope(s: HomeSummary, idx: Int): MonthScope {
    val lastIdx = maxOf(s.bookedByMonth.size - 1, 0)
    val i = idx.coerceIn(0, lastIdx)
    val currentIdx = s.currentMonthIdx.coerceIn(0, lastIdx)

    val booked = s.bookedByMonth.getOrNull(i)?.booked ?: Money.ZERO
    val orderCount = s.ordersByMonth.getOrNull(i)?.count ?: 0
    val collected = s.collectedByMonth.getOrNull(i)?.collected ?: Money.ZERO
    val paymentCount = s.collectedByMonth.getOrNull(i)?.paymentCount ?: 0

    // `i - 1`, never "last month": the pair is the selected month and its predecessor.
    val prevBooked = if (i > 0) s.bookedByMonth.getOrNull(i - 1)?.booked ?: Money.ZERO else Money.ZERO
    val prevOrderCount = if (i > 0) s.ordersByMonth.getOrNull(i - 1)?.count ?: 0 else 0
    val prevCollected = if (i > 0) s.collectedByMonth.getOrNull(i - 1)?.collected ?: Money.ZERO else Money.ZERO

    val aov = averageOrderValue(booked, orderCount)
    val prevAov = averageOrderValue(prevBooked, prevOrderCount)

    val from = maxOf(0, i - (SPARKLINE_MONTHS - 1))
    val bookedWindow = s.bookedByMonth.window(from, i) { it.booked }
    val collectedWindow = s.collectedByMonth.window(from, i) { it.collected }
    // The AOV sparkline is the headline's own formula month by month — booked ÷ orders — not
    // collections per order. A month with no orders has no average and contributes zero.
    val aovWindow = bookedWindow.mapIndexed { k, m ->
        averageOrderValue(m, s.ordersByMonth.getOrNull(from + k)?.count ?: 0)
    }

    val monthKey = s.monthKeys.getOrNull(i) ?: ""
    return MonthScope(
        idx = i,
        monthKey = monthKey,
        isCurrent = i == currentIdx,
        booked = PeriodMoney(booked, orderCount, buildTrend(booked, prevBooked, TrendPolarity.POSITIVE)),
        collected = PeriodMoney(collected, paymentCount, buildTrend(collected, prevCollected, TrendPolarity.POSITIVE)),
        // All-time is all-time: it is the one line on a rail card that does NOT follow the month.
        aov = Aov(thisMonth = aov, allTime = s.aov.allTime, trend = buildTrend(aov, prevAov, TrendPolarity.POSITIVE)),
        bookedSeries = bookedWindow,
        collectedSeries = collectedWindow,
        aovSeries = aovWindow,
        loaded = s.loadedVolumeByMonth.find { it.monthKey == monthKey },
    )
}
